toolkit/tools/pkg/imagecustomizerlib/artifactsinputoutput.go

```diff
@@ -95,15 +144,143 @@ func outputArtifacts(items []imagecustomizerapi.OutputArtifactsItemType, outputD
+	yamlBytes, err := yaml.Marshal(&yamlStruct)
+	if err != nil {
+		return fmt.Errorf("failed to marshal inject files metadata: %w", err)
+	}
+
+	// Construct YAML with header and inline unsigned comments
+	var builder strings.Builder
+	builder.WriteString("# This file is generated automatically by Prism output artifacts API.\n\n")
+
+	lines := strings.Split(string(yamlBytes), "\n")
+	i := 0
+	for _, item := range metadata {
+		for ; i < len(lines); i++ {
+			line := lines[i]
+			builder.WriteString(line + "\n")
+
+			// When you hit the 'source:' line, add the comment line
+			if strings.TrimSpace(line) == "source: "+item.Source {
+				indent := strings.Repeat(" ", leadingSpaces(line))
+				unsignedSource := replaceSuffix(item.Source, ".signed.efi", ".unsigned.efi")
+				builder.WriteString(fmt.Sprintf("%s# unsigned source: %s\n", indent, unsignedSource))
+				i++
+				break
+			}
+		}
+	}
+
+	outputFilePath := filepath.Join(baseConfigPath, "inject-files.yaml")
+	if err := os.WriteFile(outputFilePath, []byte(builder.String()), 0644); err != nil {
+		return fmt.Errorf("failed to write inject-files.yaml: %w", err)
+	}
+
+	return nil
+}
+
+func leadingSpaces(s string) int {
+	return len(s) - len(strings.TrimLeft(s, " "))
+}
+
+func (i *InjectFilesYaml) IsValid() error {
+	if len(i.InjectFiles) == 0 {
+		return fmt.Errorf("injectFiles is empty")
+	}
+	for idx, entry := range i.InjectFiles {
+		if entry.Source == "" || entry.Destination == "" {
+			return fmt.Errorf("injectFiles[%d] has empty source or destination", idx)
+		}
+		if entry.Partition.Id == "" {
+			return fmt.Errorf("injectFiles[%d] has empty partition id", idx)
+		}
+		if err := entry.Partition.MountIdType.IsValid(); err != nil {
+			return fmt.Errorf("injectFiles[%d] has invalid partition mount id type: %w", idx, err)
+		}
+	}
+	return nil
+}
+
+func InjectFiles(buildDir string, baseConfigPath string, inputImageFile string,
+	metadata []OutputArtifactMetadata,
+) error {
+	logger.Log.Debugf("Injecting Files")
+
+	loopback, err := safeloopback.NewLoopback(inputImageFile)
+	if err != nil {
+		return fmt.Errorf("failed to connect to image file to inject files:\n%w", err)
+	}
+	defer loopback.Close()
+
+	diskPartitions, err := diskutils.GetDiskPartitions(loopback.DevicePath())
 	if err != nil {
 		return err
 	}
 
+	uniqueKeys := make(map[InjectFilePartition]bool)
+	var partitions []diskutils.PartitionInfo
+
+	for _, item := range metadata {
+		key := InjectFilePartition{MountIdType: item.Partition.MountIdType, Id: item.Partition.Id}
+		if uniqueKeys[key] {
+			continue
+		}
+		uniqueKeys[key] = true
+
+		partition, _, err := findPartition(item.Partition.MountIdType, item.Partition.Id, diskPartitions, buildDir)
+		if err != nil {
+			return err
+		}
+
+		partitions = append(partitions, partition)
+		logger.Log.Infof("Identified partition %s (%s) -> device %s", item.Partition.Id, item.Partition.MountIdType, partition.Path)
+	}
+
+	// Next step: Mount & inject (to be added later)
```

Comment:
I saw partition not found issue with partuuid search, need advice here.

```
lanzeliu@fedora:~/workspace/azure-linux-image-tools/toolkit/tools$ sudo go run ./imagecustomizer \
  inject-files \
  --config-file /home/lanzeliu/workspace/test-images-uki/platform-integration-images/trident-verity-testimage/base/verity-tests/inject-files.yaml \
  --image-file /home/lanzeliu/workspace/images/safeguard/verity-root-uki-extraction.qcow2 \
  --build-dir ./build
2025/04/02 23:11:14 inject-files failed:
partition not found (part-uuid=7c74cdca-4e3e-4c3e-a78c-c2fce8eb32b9)
exit status 1
```

Reply:
Did you convert the qcow2 file to raw before mounting it as a loopback?

Reply:
Fixed, thanks.